### ISSUE TITLE
[Bug] Arguments not properly being escaped

### ISSUE THREAD

**Issue by Syphdias** [creativeprojects/resticprofile#56]
I noticed problems when using `dump` with file paths that have spaces in them.
This is not just a problem of `dump` but of all commands and also not just limited to to spaces but everything that needs to be quoted. Symbols like `'` and `"` come to mind which can be part of a file name.

This can be observed via `--dry-run` but appears that this option builds it's output independently from what is actually executed. This can also be used to execute arbitrary commands. I don't know if this would be exploitable in any way – probably not – but is definitely not how it should behave.
```
❯ resticprofile --config /tmp/profiles.toml -n local --dry-run dump 1231245125 "/path/with space; echo foo"
2021/07/23 01:56:23 profile 'local': starting 'dump'
2021/07/23 01:56:23 dry-run: /usr/bin/restic dump --password-file /home/restic/restic.local.password --repo /restic 1231245125 /path/with space; echo foo
2021/07/23 01:56:23 profile 'local': finished 'dump'
❯ resticprofile --config /tmp/profiles.toml -n local dump 1231245125 "/path/with space; echo foo"
2021/07/23 01:56:27 profile 'local': starting 'dump'
Fatal: no file and no snapshot ID specified
foo
2021/07/23 01:56:27 profile 'local': finished 'dump'
```

I think I could track the bug down to [`shell/command.go` line 184](https://github.com/creativeprojects/resticprofile/blob/f58116ba38473d1e702c55423f03f223626ea829/shell/command.go#L182):
```golang
    return shell, []string{"-c", strings.Join(flatCommand, " ")}, nil
```

I don't quite get the need for why it needs to be called with `sh -c` and not just directly via the first argument in `flatCommand`. Compatibility? Path variable? Something like this should do:
```golang
    return flatCommand[0], flatCommand[1:], nil
```
Also the dry run should use this code instead of building it's own interpretation of the arguments.

**Comment by creativeprojects** [comment]
There's a simple reason why I had to use `sh -c "restic ..."`:
The expectation of resticprofile is **not** to change the way we would interact with restic. Imagine you would normally run this kind of command:
```
$ restic backup /data*
```

The shell will expand the path before executing the command, and restic will receive (as an example) `restic backup /data0 /data1 /data2`

To be able to have this path expansion, I need to call restic through the shell. And it is not particularly convenient.

The other option would be to do the path expansion myself (and maybe some other things the shell would do to the command line). But I'm not entirely sure of the complexity of it.

**Comment by creativeprojects** [comment]
So I was looking at it in more details, what you get from using the shell are:
- environment variables
- file expansion
- pipelines
- redirection (input / output)

So we're not using the pipelines, the input/output redirection and environment variables are already managed in resticprofile. So that leaves the file expansion, which could be dealt with the filepath.Glob function.

I can try this way with an opt-in flag, not to break anything already working.

**Comment by Syphdias** [comment]
> The expectation of resticprofile is **not** to change the way we would interact with restic. Imagine you would normally run this kind of command:

That was my expectation as well `restic dump "/path/with space"` should work as well as `resticprofile -n profile dump "/path/with space"`.

> To be able to have this path expansion, I need to call restic through the shell. And it is not particularly convenient.

I think this shows that there are different kinds of arguments.

For example, if you were to run `resticprofile -n profile backup /data*` your shell would expand the glob first before passing the result to `resticprofile` (unless you quote it) – no need to handle it internally.

The second type of argument is the one taken from configuration like `source` and `exclude` (maybe more?). These ones are being escaped by `resticprofile` before passing them to `restic.`. (I think`restic` handles it internally? Otherwise my excludes are not working :sweat_smile:)

If I quote an argument (`'/data*'` or `'/path/with space'`), `restic` takes it as is and as one argument but `resticprofile` handles it differently (globbing for `'/data*'` and splitting into two variables for `'/path/with space'`).

I think we are on the same page about how it should behave. I am just not sure if I have convinced you that the current state is not the correct one :smile:

**Comment by Syphdias** [comment]
Notes and research

Places where globbing could be expanded:
1. Executing environment (local shell, systemd. cron, etc.)
2. resticprofile argument handling
3. restic argument handling (I think globbing is only expanded for excludes)

Testing 3:
```
❯ tree /tmp/baz/
/tmp/baz/
├── bar
│   └── foobar
└── foo
    └── foobar
❯ # restic backup '/tmp/baz/'
❯ restic ls 4ff5720c
snapshot 4ff5720c of [/tmp/baz] filtered by [] at 2021-07-24 15:23:03.851362169 +0200 CEST):
/tmp
/tmp/baz
/tmp/baz/bar
/tmp/baz/bar/foobar
/tmp/baz/foo
/tmp/baz/foo/foobar
❯ # restic backup /tmp/baz/* # shell globbing
❯ restic ls f53c271f
snapshot f53c271f of [/tmp/baz/bar /tmp/baz/foo] filtered by [] at 2021-07-24 15:23:26.663001485 +0200 CEST):
/tmp
/tmp/baz
/tmp/baz/bar
/tmp/baz/bar/foobar
/tmp/baz/foo
/tmp/baz/foo/foobar
❯ restic backup '/tmp/baz/*' # restic does not expand glob
/tmp/baz/* does not exist, skipping
Fatal: all target directories/files do not exist
❯ # restic backup '/tmp/baz/' --exclude '/tmp/baz/*'
❯ restic ls bca42203
snapshot bca42203 of [/tmp/baz] filtered by [] at 2021-07-24 15:24:05.710166878 +0200 CEST):
/tmp
/tmp/baz
❯ # restic backup '/tmp/baz/' --exclude /tmp/baz/* # WRONG cli usage! excludes /tmp/baz/bar and _includes_ /tmp/baz/foo
❯ restic ls becdbf0d
snapshot becdbf0d of [/tmp/baz /tmp/baz/foo] filtered by [] at 2021-07-24 15:31:20.81429059 +0200 CEST):
/tmp
/tmp/baz
/tmp/baz/foo
/tmp/baz/foo/foobar
❯ restic dump f53c271f '/tmp/baz/bar/fooba*'
cannot dump file: path "/tmp/baz/bar/fooba*" not found in snapshot
```

Conclusion for globbing in `restic`: Not for sources, not for dump, only for excludes.

**Comment by creativeprojects** [comment]
> I think we are on the same page about how it should behave. I am just not sure if I have convinced you that the current state is not the correct one 😄

I agree something is wrong the way it is now.
Also the leftover on the command line shouldn't be managed in a different way.
I need to come up with a solution dealing with both kind of parameters in the same way.

I'll still need to pass it through the shell though, because the `filepath.Glob` function is not dealing with a `~` in the path (and maybe some other esoteric things I'm forgetting)

**Comment by Syphdias** [comment]
What do you mean by "leftover on the command line"?

> I'll still need to pass it through the shell though, because the `filepath.Glob` function is not dealing with a `~` in the path (and maybe some other esoteric things I'm forgetting)

I will be mean and play QA here: What if you want to backup an actual folder named `~` :grimacing:.
Again this might be something that should be handled by the shell before passing it as argument.

**Comment by creativeprojects** [comment]
> What do you mean by "leftover on the command line"?

I mean the parameters coming from the resticprofile command line, not the configuration file (typically the examples with the `dump` command)

> I will be mean and play QA here: What if you want to backup an actual folder named `~` 😬.

You can but you either need a fixed path, like `./~` or escape it `\~`

**Comment by djmattyg007** [comment]
> Alternatively, there's something I wanted to look at for a while, which is to try to use the restic code as a library and get it integrated into resticprofile. But there's a lot of work on that!

You'll need to be careful doing this. restic is a BSD-licensed codebase, while yours is a GPL-licensed codebase.

**Comment by creativeprojects** [comment]
> You'll need to be careful doing this. restic is a BSD-licensed codebase, while yours is a GPL-licensed codebase.

Actually I don't think I'll go this path.
I looked at it again and it will introduce a few other issues (things like a `run-on-failure` might no longer work depending on how restic fails)

I think it's still better to use restic as a command line.

Anyway I got something working nicely 👍🏻

**Comment by creativeprojects** [comment]
This time I'm going to get this right. So there are two ways of escaping:
- a string that allows shell glob (`*`, `?` and `[`)
- a string that does **not** allow shell glob (for `exclude` and `exclude`)

In the first case, all the special characters are escaped and the string is **not** quoted.
In the second case, the string is in double quotes if it has any special character.

Please note I'm not making use of the single quote at all.

Does that sound right?

Here's the output of a unit test checking both methods:

   |Argument        |Shell escape with glob   |Not allowing Glob   |
   | --- | --- | --- |
   |simple                       |simple                               |simple                           |
   |Côte d'Ivoire                |Côte\\ d\\'Ivoire                    |\"Côte d'Ivoire\"                |
   |quo\"te                      |quo\\\"te                            |\"quo\\\"te\"                    |
   |quo'te                       |quo\\'te                             |\"quo'te\"                       |
   |with space                   |with\\ space                         |\"with space\"                   |
   |quo\\\"te                    |quo\\\"te                            |quo\\\"te                        |
   |quo\\'te                     |quo\\'te                             |quo\\'te                         |
   |/path/with space; echo foo   |/path/with\\ space\\;\\ echo\\ foo   |\"/path/with space; echo foo\"   |
   |**/.git/                     |**/.git/                             |\"**/.git/\"                     |
   |[aA]*                        |[aA]*                                |\"[aA]*\"                        |

Do you think of any other funny input for the unit test?

**Comment by Syphdias** [comment]
> Do you think of any other funny input for the unit test?

As you mentioned yourself. Maybe include something with `?`. Also maybe something with a `$`.

| Argument | Shell escape with glob | Not allowing Glob |
|-|-|-| 
| /etc/p?? | /etc/p\\?\\? | "/etc/p??" |
| $foo | \\$foo | '$foo' |

> Please note I'm not making use of the single quote at all.

Mind the single quotes for the variable one.

For zsh at least there is more characters used for globs, but that does not matter.

**Comment by creativeprojects** [comment]
Fix released in v0.15.0
